Switch safekeepers to async.

arssher opened the pull request:
This is a full switch, fs io operations are also tokio ones, working through thread pool. Similar to pageserver, we have multiple runtimes for easier `top` usage and isolation.

Notable points:
- Now that guts of safekeeper.rs are full of .await's, we need to be very careful not to drop task at random point, leaving timeline in unclear state. Currently the only writer is walreceiver and we don't have top level cancellation there, so we are good. But to be safe probably we should add a fuse panicking if task is being dropped while operation on a timeline is in progress.
- Timeline lock is Tokio one now, as we do disk IO under it.
- Collecting metrics got a crutch: since prometheus Collector is synchronous, there is now a special task copying once in a scrape period data from under async lock to sync one where collector can take it.
- Anything involving closures becomes significantly more complicated, as async fns are already kinda closures + 'async closures are unstable'.
- Main thread now tracks other main tasks, which got much easier.
- The only sync place left is initial data loading, as otherwise clippy complains on timeline map lock being held across await points -- which is not bad here as it happens only in single threaded runtime of main thread. But having it sync doesn't hurt either.

I'm concerned about performance of thread pool io offloading, async traits and many await points; but we can try and see how it goes.

fixes https://github.com/neondatabase/neon/issues/3036 fixes https://github.com/neondatabase/neon/issues/3966

arssher commented:
As a basic perf test, we can try pgbench -i on staging before and after this patch.

arssher commented:
Since metrics now can be delayed up to a scrape period due to clutch described above, tests relying on them fail.

petuhovskiy commented:
Delaying metrics doesn't feel right, we cannot rely on them now because they can be delayed. Can we .await timeline metrics to make it work like it was?

petuhovskiy reviewed (commented):
Generally looks ok, but I wouldn't want to roll it out to production until we test it for at least several days.

It's hard to say if we forgot/missed something, my main concerns:
- blocking calls from async tasks, potential thread pool exhaustion
- forgetting to .await (but AFAIR compiler checks for it?)
- other bugs or changed behavior

Would be good to have metrics for async runtimes, can we somehow add that? The most useful metrics are:
- number of unfinished spawned tasks
- number of active (non-awaiting IO) tasks
- total number of spawned tasks

koivunej commented on safekeeper/src/bin/safekeeper.rs:
```diff
@@ -220,71 +230,77 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
+        _ = sigterm_stream.recv() => info!("received SIGTERM, terminating")
+
+    };
+    std::process::exit(0);
```
This is simply redundant because main `await`'s on this fn.

koivunej replied:
Resolving this in favor of my other thread which discusses more of the bad effects of not using `std::process::exit` instead of running all drops.

koivunej commented on safekeeper/src/bin/safekeeper.rs:
```diff
@@ -220,71 +230,77 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
-                .unwrap();
-            })?,
-    );
+    // Keep handles to main tasks to die if any of them disappears. Probably
+    // replace it with JoinSet once we update tokio.
```
What does this comment mean? `safekeeper` is already using tokio 1.27.0 like rest of the workspace. The minimum version in our `Cargo.toml` is not the real minimum because we have 1.27.0 in `Cargo.lock`.

Very much agree on the joinset usage, however it will not allow you to do the similar `(String, Result<Result<_, _>, JoinError>)` with `futures::future::FutureExt::map`... Perhaps you were referring to the `JoinSet::build_task(&mut self)` and `Builder::name` being unstable?

Alternative to this would be to have just plain futures here locally (each as they are `tokio::spawn(...).map(...)`) and then `tokio::try_join!(...)` over all of them. The `FuturesUnordered` provides very little.

arssher replied:
> The minimum version in our Cargo.toml is not the real minimum because we have 1.27.0 in Cargo.lock.

Oh, that was confusing, thanks. No, I meant specifically JoinSet.

So I think FuturesUnordered is ok here, removed comment.

koivunej commented on safekeeper/src/bin/safekeeper.rs:
```diff
@@ -220,71 +230,77 @@ fn start_safekeeper(conf: SafeKeeperConf) -> Result<()> {
+    tokio::select! {
+        Some((task_name, res)) = tasks_handles.next()=> {
+            error!("{} task failed: {:?}, exiting", task_name, res);
+            std::process::exit(1);
```
This will skip all of the drop handlers, which might be surprising while trying to debug.

Would recommend either returning a exit code from this fn or an `Error` to be logged by `main`.

With returning, it's good to keep in mind that by using `FuturesUnordered` or anything else than a `JoinSet`, you will only see a random first error here but the spawned tasks will continue to run until in the background regardless of exiting this method.

Still, I think it would be most sensible to allow dropping or dtors to run. Because if someone adds a dropguard anywhere, and it doesn't run because of some other task failed and bubbled up to these exits, you will have zero logging on that and no way to confirm it running or not.

koivunej replied:
Again I initially missed that `FuturesUnordered` is used to turn every `Ok(())` into an `Err`, so try_join! will not work without additional massaging.

koivunej replied:
Cannot see a way to do this with joinset, but here's the try_join idea: https://github.com/neondatabase/neon/pull/4124/commits/4d96a27e3ec5f4e933c1482e0dc65fc7ed24d31c

arssher replied:
Safekeepers are not the complicated. There is no graceful shutdown of any kinds and no drops which are expected to be run on shutdown, and I prefer to keep things simple unless otherwise is required, so this seems to be a bit premature.

koivunej replied:
This is different from graceful shutdown, it's about "clean" as expected in rust applications shutdown, so that destructors are run (when not doing abort). I don't know if it was in place before, but now would be a good time to add it. It might expose some deadlocks at worst but systemd will send a SIGKILL eventually.

arssher replied:
But do you have any idea of concrete examples where this would be needed?

This is one of the areas where I personally much more bothered by other things, but don't mind someone picking it up.

arssher replied:
btw, I don't see what try_join changes here; stopping polling handles doesn't mean tasks are stopped. Doing return instead of std::process::exit is possible with older version as well. Not sure what are rust guarantees when main thread finishes; will it call drops in other threads? Probably yes.

koivunej replied:
Aah yes, actually dropping with multiple runtimes... I don't know what happens. The runtime running `async fn main` will get dropped at least.

But, ok. Don't want to hold this.

koivunej commented on safekeeper/src/wal_service.rs:
```diff
@@ -16,104 +16,82 @@ use crate::SafeKeeperConf;
-                }
+        tokio::spawn(async move {
+            if let Err(err) = handle_socket(socket, conf, conn_id)
+                .instrument(info_span!("", cid = %conn_id))
```
This looks like a mistake, why not?
```suggestion
                .instrument(info_span!("wal service", cid = %conn_id))
```

However usually it is more ergonomic to add these at the method/function level with `#[instrument(...)]`.

koivunej replied:
After looking at bit of safekeeper logs, perhaps this is by design and wouldn't help much.

koivunej commented on safekeeper/src/lib.rs:
```diff
@@ -91,11 +91,62 @@ impl SafeKeeperConf {
     }
 }
+
+// Tokio runtimes.
```
Why add all these runtimes?

One would be enough. This combination also has maximum thread count of 6 * 512 assuming they use `spawn_blocking` directly or indirectly. A single runtime should be very much enough if all file access apart from initial loading is done with tokio's functionality.

A single runtime would allow testing with a single thread runtime, which is an easy way to spot any deadlocks in a system. For this huge amount of worker threads it will be very difficult.

koivunej replied:
Looks like the `async fn main` is already driven by a `current_thread` runtime, which is great. If all of these were to be removed, and only `tokio::spawn` used, you'd get the deadlock "testing" right away.

koivunej replied:
https://github.com/neondatabase/neon/pull/4124/commits/ed4fd15b7cb122f5232787b8a8750e594e84a699 has two test failures, unsure yet if they are transient.

petuhovskiy replied:
Those failures are "expected":

> Since metrics now can be delayed up to a scrape period due to clutch described above, tests relying on them fail.

Not sure yet what is the good fix for the metrics.

koivunej reviewed (changes requested):
Few notes and questions on the not running drops on error and how the tasks are currently launched.

On a high level I still find `GlobalTimelinesState` behind an std mutex. Is that reasonable global std mutex? If we don't have time to do proper analysis, would recommend just going with tokio's async mutex. EDIT: it seems that it is required by the current way of doing metrics via the collector api.

Requesting changes after seeing the 6 tokio runtimes: I don't see any justification and one should be enough.

Looks like there is no `spawn_blocking` usage in the codebase which removes one deadlock source. Similarly no more `block_on` or `block_in_place` which is awesome :+1:.

koivunej commented:
> forgetting to .await (but AFAIR compiler checks for it?)

Future trait has a `#[must_use = "futures do nothing unless you `.await` or poll them"]` and similar have been added to tokio for named futures.  In general if a future type does not have a `#[must_use]` then it is a bug.. Can't say I've ran into this in other codebases; we did have a `let _ = future_which_was_never_awaited;` in the previous iteration of `postgres_backend`. It was caught by upgrading clippy.

petuhovskiy commented:
> On a high level I still find `GlobalTimelinesState` behind an std mutex. Is that reasonable global std mutex? If we don't have time to do proper analysis, would recommend just going with tokio's async mutex.

It's ok behind an std mutex. This map is just a collection of `Arc`s, we don't do anything but `get_all/get/set/delete` while holding a lock, so it should never block anyone.

petuhovskiy commented:
> Currently the only writer is walreceiver and we don't have top level cancellation there, so we are good.

HTTP handlers can and do get canceled often, need to check that we're good there.

koivunej commented:
> > Currently the only writer is walreceiver and we don't have top level cancellation there, so we are good.
> 
> HTTP handlers can and do get canceled often, need to check that we're good there.

An obvious way would be to spawn on each ... At the same time then I think you'd need a way to coalesce the requests, otherwise later attempts would get confusing or contradictory responses.

I see that you are not yet using `RequestSpan` around your handlers. Is this by design? It will at least *log* if request is cancelled before completion.

petuhovskiy commented:
We miss logging on the HTTP side, it would be good to add spans and log at least some requests.

koivunej commented:
> We miss logging on the HTTP side, it would be good to add spans and log at least some requests.

Unless I am missing something, this will be it: https://github.com/neondatabase/neon/pull/4124/commits/ab0ecf55489390f47a4ba675c08a138e85cb89a8

With the panic tracing hook any panic will be within the correct context because that is before unwinding, and things will just work.

Noticed that on panic, the cancellation message will be printed *but* it's not that big of a deal. I'll PR that fix separatedly.

arssher commented:
> Delaying metrics doesn't feel right, we cannot rely on them now because they can be delayed. Can we .await timeline metrics to make it work like it was?

Came up with another version of crutch, now they are fresh and tests are fixed.

arssher commented:
> On a high level I still find GlobalTimelinesState behind an std mutex. Is that reasonable global std mutex? If we don't have time to do proper analysis, 

We do; as Arthur said, we only hold it during small pieces of CPU job and never await while holding, so std is fine.

arssher commented:
> Requesting changes after seeing the 6 tokio runtimes: I don't see any justification and one should be enough.

Having multiple runtimes provides isolation. For example, it gives immediate breakdown by subsystems when you look at top. Also, imagine there is a bug in http processing code which blocks the runtime in some blocking operations; it would be probably good if that doesn't harm WAL servicing. Yes, clearly overall we get more runtime threads than needed. But I don't expect that having 2x or 3x number of cores threads is sensibly worse than 1x from perf POV, though I'm not an expert here. So I'd leave it as is, though can live with either variant.

arssher commented:
> Unless I am missing something, this will be it: https://github.com/neondatabase/neon/commit/ab0ecf55489390f47a4ba675c08a138e85cb89a8

Great, included it in this PR.

arssher commented:
> Would be good to have metrics for async runtimes, can we somehow add that? The most useful metrics are:

Yeah. So there is [unstable TODO metrics](https://docs.rs/tokio/latest/tokio/runtime/struct.Runtime.html#method.metrics) in tokio, and [tokio-metrics](https://github.com/tokio-rs/tokio-metrics), also requiring tokio unstable. So no good solution off the shelf. Also we can always enable/attach tokio console and check out what's going on.

koivunej commented:
The easiest way to get metrics today is to add future wrappers to spawned futures before boxing, count polls etc like tokio-console.

Note one cannot simply attach a tokio console, it will now require rebuilding with `--cfg tokio_unstable` unless it has changed recently.

vadim2404 commented on safekeeper/src/bin/safekeeper.rs:
```diff
@@ -120,7 +122,8 @@ struct Args {
 }
 
-fn main() -> anyhow::Result<()> {
+#[tokio::main(flavor = "current_thread")]
```
Why do you need only single-threaded env?

arssher replied:
Because this is for the main task, which does almost nothing.

arssher commented:
I rebased and added option to run everything in single threaded runtime (for debugging etc). Will try to run tests and commit today.

arssher commented:
Before merging I wanted to run at least basic benchmarks to have a feeling of thread pooled fs io operations overhead. I did that with pgbench insertion. In short, results are that overhead is noticeable, but can be neglected at least for now.

The command was like
```
pgbench -d  'postgres://ars:qB5QsTIaX0Sz@ep-dry-hill-850843.eu-west-1.aws.neon.build/neondb' -i -s 600 -I dtG
```
On eu-west staging, run on the ec2 machine in the same region:
sync:
server-side generate 71.91 s, disk write stable 261MiB/s.
server-side generate 71.96 s 224MiB/s
server-side generate 72.11 s 176MiB/s.
server-side generate 72.07 s 234MiB/s

async:
server-side generate 71.97 s disk load 198MiB/s.
server-side generate 72.18 s 262MiB/s.
server-side generate 72.09 s 263MiB/s.

async, current thread rt:
server-side generate 72.25 s 196MiB/s.
server-side generate 72.13 s 251MiB/s

Scale 600 is about 9.5GiB of WAL. Disk throughput is taken from grafana (node_exporter, which in turn reads /proc counters). I'm not fully sure why it has such variance, but probably 1) test is just too short (~70s, but grafana scrapes each 30s) 2) WAL offloading / sending to pageserver takes some of disk capacity, and this is not stable due to s3 throttling.

As kinda expected, single thread runtime is as good as multi as technically network reading and disk IO still happen in different threads due to thread pool offload.

Running the same pgbench but locally, with scale 200 now, having everything (compute, single safekeeper, single pageserver) in tmpfs:
async:
server_side_generate: 22.210 s
server_side_generate: 22.260 s
server_side_generate: 19.130 s
server_side_generate: 18.830 s
server_side_generate: 19.450 s
sync:
server_side_generate: 19.130 s
server_side_generate: 17.140 s
server_side_generate: 17.620 s
server_side_generate: 17.490 s
server_side_generate: 18.080 s
So it seems to be about ~10% difference.
command was like

```
BUILD_TYPE=release CARGO_BUILD_FLAGS="--features=testing" make -s -j12 && rm -rf /tmp/test_output/; TEST_OUTPUT=/tmp/test_output NEON_BIN=./target/release poetry run pytest test_runner/performance -k test_perf_pgbench --preserve-database-files
```
but I also adjusted code a bit (left only init test with -I dtG steps, removed vanilla, changed scale).

Importantly, both on staging and locally I disabled backpressure
```
+            "max_replication_apply_lag=-1",
+            "max_replication_flush_lag=-1",
+            "max_replication_write_lag=-1"
```

otherwise it kicks in quickly and variance increases a lot. Actually, before I realized that I need to do that I run some tests and it seems like there sync vs async diff was greater -- I remember at least one run with async taking 38.4s and sync 32s, but not sure how reliable is that.

While doing that, I found multiple interesting and unresolved links/issues about bad fs io thread pool performance:
https://users.rust-lang.org/t/why-tokio-fs-is-slower-than-std-fs/39610
https://users.rust-lang.org/t/blocking-permit/36865 (old but referenced fresh link above)
https://github.com/tokio-rs/tokio/issues/3664
They seem to be more about listing operations though.
Overall, it looks good enough currently, so I merged the PR.
